servers/zts/src/main/java/com/yahoo/athenz/zts/ZTSImpl.java

```diff
@@ -2283,21 +2282,135 @@ public RoleCertificate postRoleCertificateRequestExt(ResourceContext ctx, RoleCe
+
+           if (!workloadList.isEmpty()) {
+               workloads.getWorkloadList().addAll(workloadList);
+           }
+        }
+
+        //get dynamic list from workloads store
+        List<Workload> dynamicWorkloadList = instanceCertManager.getWorkloadsByService(finalDomainName, finalServiceName);
+
+        workloads.getWorkloadList().addAll(dynamicWorkloadList);
+        return workloads;
     }
 
     @Override
-    public Workloads getWorkloadsByIP(ResourceContext context, String ip) {
-        // to be implemented in next PR
-        return null;
+    public Workloads getWorkloadsByIP(ResourceContext ctx, String ip) {
+
+        final String caller = ctx.getApiName();
+        final String principalDomain = logPrincipalAndGetDomain(ctx);
+
+        if (readOnlyMode) {
+            throw requestError("Server in Maintenance Read-Only mode. Please try your request later",
+                    caller, ZTSConsts.ZTS_UNKNOWN_DOMAIN, principalDomain);
+        }
+
+        validateRequest(ctx.request(), principalDomain, caller);
+        validateIpAddress(ip, "getWorkloadsByIP", null, principalDomain);
+
+        Workloads workloads = new Workloads();
+        workloads.setWorkloadList(new ArrayList<>());
+
+        //get dynamic list from workloads store
+        List<Workload> dynamicWorkloadList = instanceCertManager.getWorkloadsByIp(ip);
+
+        workloads.getWorkloadList().addAll(dynamicWorkloadList);
+        return workloads;
     }
 
     @Override
-    public TransportRules getTransportRules(ResourceContext context, String domainName, String serviceName) {
-        // to be implemented in next PR
-        return null;
+    public TransportRules getTransportRules(ResourceContext ctx, String domainName, String serviceName) {
+        final String caller = ctx.getApiName();
+        final String principalDomain = logPrincipalAndGetDomain(ctx);
+
+        if (readOnlyMode) {
+            throw requestError("Server in Maintenance Read-Only mode. Please try your request later",
+                    caller, ZTSConsts.ZTS_UNKNOWN_DOMAIN, principalDomain);
+        }
+
+        validateRequest(ctx.request(), principalDomain, caller);
+        validate(domainName, TYPE_DOMAIN_NAME, principalDomain, caller);
+        validate(serviceName, TYPE_ENTITY_NAME, principalDomain, caller);
+
+        TransportRules transportRules = new TransportRules();
+        transportRules.setIngressRules(new ArrayList<>());
+        transportRules.setEgressRules(new ArrayList<>());
+
+        Map<String, List<String>> transportRulesBaseData = dataStore.getDataCache(domainName).getTransportRulesInfoForService(serviceName);
+        Workloads workloads;
+        if (transportRulesBaseData != null) {
+            for (Map.Entry<String, List<String>> entry : transportRulesBaseData.entrySet()) {
+                TransportRulesProcessor.TransportDirection transportDirection = TransportRulesProcessor.getDirection(entry.getKey());
```

Comment:
why do we have 2 separate lines 2387-2388? calling this twice requires the server to run the regex match twice. Shouldn't the transportDirection be part of the TransportRule?